### ISSUE TITLE
When running multiple test methods the first method finishes and the second one stucks

### ISSUE THREAD

**Issue by eklierka** [spring-projects/spring-kafka#653]
I send a message from transactional Kafka listener to have message sending and offset committing in the same transaction. 

**TestKafkaListener**

```java
@Component
public class TestKafkaListener {

  private static final Logger LOGGER = LoggerFactory.getLogger(TestKafkaListener.class);

  public static final String INPUT_TEST_TOPIC = "transaction-sample-topic-in";
  public static final String OUTPUT_TEST_TOPIC = "transaction-sample-topic-out";

  @Autowired
  private KafkaTemplate<String, String> kafkaTemplate;

  @KafkaListener(topics = INPUT_TEST_TOPIC)
  @Transactional("kafkaTransactionManager")
  public void listen(ConsumerRecord<String, String> record) {
    LOGGER.info("Received Kafka record from {}: {}", INPUT_TEST_TOPIC, record);
    kafkaTemplate.send(OUTPUT_TEST_TOPIC, record.key(), record.value());
    LOGGER.info("Forwarded Kafka record to {}: {}", OUTPUT_TEST_TOPIC, record);
    stubMethod(record.value());
  }

  private void stubMethod(String data) {
    if (data.equals("invalid_data")) {
      throw new RuntimeException("Invalid data");
    }
  }
}
```

**application.properties**

```properties
spring.kafka.consumer.auto-offset-reset=earliest
spring.kafka.consumer.enable-auto-commit=false
spring.kafka.consumer.group-id=transaction-sample
spring.kafka.listener.ack-mode=RECORD

spring.kafka.producer.transaction-id-prefix=transaction-sample-${random.uuid}
```

I tried to run the following test class but after one test method finishes the second one stucks.

**KafkaListenerTest**

```java
@TestPropertySource("classpath:test.properties")
@RunWith(SpringRunner.class)
@SpringBootTest
public class KafkaListenerTest {

  @Autowired
  private KafkaEmbedded kafkaEmbedded;

  @Autowired
  private KafkaTemplate<String, String> kafkaTemplate;

  @Test
  public void shouldProcessEvent() throws Exception {
    String testKey = "test_key";
    String testData = "test_data";

    kafkaTemplate.executeInTransaction(kt ->
        kt.send(INPUT_TEST_TOPIC, testKey, testData));

    try (Consumer<String, String> consumer = createConsumer()) {
      kafkaEmbedded.consumeFromAnEmbeddedTopic(consumer, OUTPUT_TEST_TOPIC);
      ConsumerRecords<String, String> records = KafkaTestUtils.getRecords(consumer);
      Iterator<ConsumerRecord<String, String>> iterator = records.iterator();
      ConsumerRecord<String, String> record = iterator.next();

      assertEquals(testKey, record.key());
      assertEquals(testData, record.value());
      assertFalse(iterator.hasNext());
    }
  }

  private Consumer<String, String> createConsumer() {
    Map<String, Object> consumerProps =
        KafkaTestUtils.consumerProps("test-consumer", "true", kafkaEmbedded);
    consumerProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
    consumerProps.put(ConsumerConfig.ISOLATION_LEVEL_CONFIG, "read_committed");
    DefaultKafkaConsumerFactory<String, String> cf = new DefaultKafkaConsumerFactory<>(
        consumerProps, new StringDeserializer(), new StringDeserializer());
    return cf.createConsumer();
  }

  @Test
  public void shouldRollback() throws Exception {
    String testKey = "test_key";
    String testData = "invalid_data";

    kafkaTemplate.executeInTransaction(kt ->
        kt.send(INPUT_TEST_TOPIC, testKey, testData));

    try (Consumer<String, String> consumer = createConsumer()) {
      kafkaEmbedded.consumeFromAnEmbeddedTopic(consumer, OUTPUT_TEST_TOPIC);
      ConsumerRecords<String, String> records = KafkaTestUtils.getRecords(consumer);

      assertFalse(records.iterator().hasNext());
    }
  }
}
```

**KafkaEmbeddedConfig**
```java
@Configuration
public class KafkaEmbeddedConfig {

  @Bean
  public KafkaEmbedded kafkaEmbedded() {
    KafkaEmbedded kafkaEmbedded = new KafkaEmbedded(1, false, 1,
        INPUT_TEST_TOPIC, OUTPUT_TEST_TOPIC);
    kafkaEmbedded.setKafkaPorts(9092);
    kafkaEmbedded.brokerProperty(KafkaConfig.TransactionsTopicReplicationFactorProp(), "1");
    kafkaEmbedded.brokerProperty(KafkaConfig.TransactionsTopicMinISRProp(), "1");
    return kafkaEmbedded;
  }
}
```
Is there something wrong?

I've created a sample application to reproduce the problem: https://github.com/eklierka/sample

**Comment by garyrussell** [comment]
What do you mean by "stucks"? Where is is stuck? Have you taken a thread dump.

`KafkaTestUtils.getRecords(consumer)` is not intended for testing that "no" message is received - it will block for 60 seconds by default.

**Comment by garyrussell** [comment]
BTW, you shouldn't have `@Transactional` on your listener. Instead, inject the transaction manager into the listener container. That way, the container can send the offset(s) to the transaction.

**Comment by eklierka** [comment]
I discovered that when running on Linux everything is fine while when running on Windows transactions behavior is strange. 

Now I've got a different question.
After transaction rollback the message is consumed from the topic again. The process of consuming the message and rolling back the transaction continues in indefinite loop. If I provide error handler for container then exception isn't thrown and transaction isn't rolled back. Can I provide error handler (for example to send message to DLQ) and roll back the transaction at the same time?
